src/daemon/daemon.cpp

```diff
@@ -2839,3 +2906,244 @@ void mp::Daemon::finish_async_operation(QFuture<AsyncOperationStatus> async_futu
+                {
+                    auto stream = MP_FILEOPS.open_read(fmt::format("{}/{}", cloud_init_mount_point, filename));
+                    auto contents = std::string{std::istreambuf_iterator{*stream}, {}};
+                    if (stream->fail())
+                        throw HyperkitMigrationRecoverableError{
+                            vm_name, fmt::format("could not read cloud-init config file '{}'", filename),
+                            std::strerror(errno)};
+
+                    qemu_iso.add_file(filename, contents);
+                }
+
+                qemu_iso.write_to(QString::fromStdString(fmt::format("{}/{}", target_directory, cloud_init_iso_name)));
+
+                // Migrate JSON for instance image
+                auto instance_image_record = hyperkit_instance_images_json[key].toObject();
+                auto image_record = instance_image_record["image"].toObject();
+                if (image_record.isEmpty()) // true if either the image or instance objects were empty/absent
+                    throw HyperkitMigrationRecoverableError{vm_name, "could not update instance image",
+                                                            "corrupted image records"};
+
+                image_record.insert("path", new_image);
+                image_record.insert("kernel_path", "");
+                image_record.insert("initrd_path", "");
+                instance_image_record.insert("image", image_record);
+                qemu_instance_images_json.insert(key, instance_image_record);
+
+                // Add JSON for QEMU instance
+                qemu_instances_json.insert(key, vm_spec_to_json(vm_specs));
+
+                [[maybe_unused]] auto succeeded = instances_migrated.insert(vm_name).second;
+                assert(succeeded);
+
+                rmdir_guard.dismiss(); // we succeeded migrating the instance, so don't roll back
+            }
+            catch (const HyperkitMigrationRecoverableError& e)
+            {
+                mpl::log(mpl::Level::error, category, e.what());
+                if (ret.ok())
+                    ret = grpc::Status{grpc::StatusCode::FAILED_PRECONDITION, "Faulty migration"};
+            }
+        }
+    }
+
+    std::string outcome_summary{"No instances were migrated."};
+    if (!instances_migrated.empty())
+    {
+        // Write QEMU instance and instance-image DBs
+        write_longer_json(*qemu_instances_db, qemu_instances_doc, qemu_instances_json);
+        write_longer_json(*qemu_instance_images_db, qemu_instance_images_doc, qemu_instance_images_json);
+
+        constexpr auto separator = "\n  ";
+        constexpr auto hint_delete =
+            "Multipass retained the original hyperkit instances, but they take space on disk. You can "
+            "delete them (multipass delete --purge <instance-name>) when you are confident that their "
+            "migrated counterparts are intact, by temporarily  moving back to hyperkit.";
```

Comment:
OK on most of it, but there are a couple of things that I think could be better: 
  1. Add "you can" just before "temporarily", to avoid giving the user an order (they can still decide to keep the originals if they want). 
  2. "Migration is successful" sounds a little vague. Users don't know what the migration process involves, so I would prefer not to ask them to decide whether it was successful in general. "Migrated counterparts are intact" is much more concrete, something the user can judge according to their needs. But it can surely be simplified and improved. How about something like "When you are happy with the new/migrated instances..."?

Reply:
Hey @ricab!

Ok, I agree with both of your points.  Thanks!

Reply:
Hey @townsend2010, but do you prefer "new" or "migrated" in that sentence? Or do you thing we should have that slash form?

Reply:
Yep, that should remove any ambiguity. Sounds good, thanks!

Reply:
Done. I also added a separate commit with a little fine-tuning: use proper case when mentioning Hyperkit and QEMU; remove "hyperkit" from "original hyperkit instances", to avoid redundancy (we mention hyperkit again later in the sentence); added "back" to "temporarily switch *back* to". Let me know if you like it.

Reply:
Great, thanks!  I like the wording better.  I'll check it one more time in a console and then hopefully this will be all good.